cycle login/logout while using in more than one topview the multiple flightpath widget

ReimarBauer opened the pull request:
first step a test which reproduces parts of issue 2037
test for login/logout issue on a combination of topviews and multiple… flightpath while logged in

**Purpose of PR?**:

Fixes #2037

ReimarBauer committed:
use a compatible py3 version for RTFD (#2187)

* imp was removed in py-3.12

matrss committed:
Remove executable bit from text files (#2180)

matrss committed:
Install pyvirtualdisplay in testing workflow (#2205)

The testing-stable container image no longer includes pyvirtualdisplay
due to a change that made it unnecessary for develop that also affected
the stable image. This installs the package as part of the testing
workflow instead. This is a band-aid to fix the stable CI until all of
the recent changes to the tests are backported from develop to stable.

rohit2p committed:
workflow_dispatch: trigger is added (#2197)

Co-authored-by: ReimarBauer <rb.proj@gmail.com>

ReimarBauer committed:
any member of an operation should see who created the operation (#2216)

* enables to see from any operations user the creator

* test and changes to allow for any operations member to get the creators name

* improved

ReimarBauer committed:
separate urls and menu to server parts (#2211)

* extracted from index.py

* refactored index.py, moved server dependent parts to theire modules

* updated

ReimarBauer committed:
catch exception when USE_SAML2 is True but configuration not completed (#2214)

* catch exception

* improved

* Update mslib/mscolab/server.py

Co-authored-by: Matthias Riße <9308656+matrss@users.noreply.github.com>

---------

Co-authored-by: Matthias Riße <9308656+matrss@users.noreply.github.com>

ReimarBauer committed:
test for login/logout issue on a combination of topviews and multiple flightpath while logged in

ReimarBauer committed:
A disconnect can be done only once, regardless of the docking widget usage frequency.

matrss commented on mslib/msui/topview.py:
```diff
@@ -379,12 +379,16 @@ def openTool(self, index):
-        self.mainwindow_signal_activate_operation.disconnect()
-        self.mainwindow_signal_permission_revoked.disconnect()
-        self.mainwindow_signal_render_new_permission.disconnect()
+        # A disconnect can be performed only once.
+        try:
+            self.mainwindow_signal_login_mscolab.disconnect()
+            self.mainwindow_signal_logout_mscolab.disconnect()
+            self.mainwindow_signal_listFlighttrack_doubleClicked.disconnect()
+            self.mainwindow_signal_activate_operation.disconnect()
+            self.mainwindow_signal_permission_revoked.disconnect()
+            self.mainwindow_signal_render_new_permission.disconnect()
+        except TypeError as ex:
+            logging.debug("mscolab connection already closed: %s", ex)
```
These `.disconnect()` calls disconnect *all* slots that are connected to the respective signals, which belong to a different object. I.e. when the topview window is closed all slots, even from different windows, that are connected to the mainwindow's signals get disconnected. This seems wrong to me. What is the intention of this code?

ReimarBauer replied:
when you use the multiple flightpath docking widget with mscolab, you have one connect done  to one server. There you use then different operations over this connection.

At the moment the logout in the profile is used, a disconnect to the server is driven and this can only be done once.

Without this exception handler we get a problem in the teardown which looks like this.

```
_ ERROR at teardown of Test_Mscolab.test_multiple_views_and_multiple_flightpath _
INFO:root:disconnected
INFO:root:dasdas
127.0.0.1 - - [19/Feb/2024 20:25:51] "GET /socket.io/?transport=websocket&EIO=4&sid=-blb&t=1708370751.269579 HTTP/1.1" 200 0 0.700590
INFO:root:Already disconnected from the server and switched to Fligh Tracks: True
INFO:root:Already disconnected from the server and switched to Fligh Tracks: True
DEBUG:root:switch_to_local
DEBUG:root:update_active_flight_track
Exceptions caught in Qt event loop:
________________________________________________________________________________
Traceback (most recent call last):
  File "/home/reimar/PycharmProjects/MSS/mslib/msui/topview.py", line 383, in closed
    self.mainwindow_signal_login_mscolab.disconnect()
TypeError: disconnect() failed between 'signal_login_mscolab' and all its connections

```
So yes it may hide a different problem in the teardown.  For the multiflight docking widget in a real example it is maybe not needed.

With the section removed the example shows that the second change may be enough:

![without_topview](https://github.com/Open-MSS/MSS/assets/278728/05f8f8ea-cc33-4db0-aa22-79c0d41a3b12)

Output:
```
INFO: MSS Version: 8.3.3
INFO: Python Version: 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:23:14) [GCC 10.4.0]
INFO: Platform: Linux-6.5.0-10022-tuxedo-x86_64-with-glibc2.35 (('64bit', 'ELF'))
INFO: Already disconnected from the server and switched to Fligh Tracks: True
INFO: Already disconnected from the server and switched to Fligh Tracks: True
```

ReimarBauer replied:
the function is not meant for closing the topview.

matrss replied:
> With the section removed the example shows that the second change may be enough:

Do I understand that right, that the entire method can be removed and the test passes?

> the function is not meant for closing the topview.

Right, I miss-interpreted its name. It's called when the parent view closes, at which point it also makes little sense to disconnect all slots connected to signals on the mscolab mainwindow, IMO.

ReimarBauer replied:
_Do I understand that right, that the entire method can be removed and the test passes?_

good idea, I try later.  I meant when I remove my changes, I can create the peek gifcycle but the test fails.

matrss commented on mslib/msui/multiple_flightpath_dockwidget.py:
```diff
@@ -177,16 +178,21 @@ def __init__(self, parent=None, view=None, listFlightTracks=None,
-        for idx in range(len(self.obb)):
-            del self.obb[idx]
+        try:
+            self.ui.signal_listFlighttrack_doubleClicked.disconnect()
+            self.ui.signal_permission_revoked.disconnect()
+            self.ui.signal_render_new_permission.disconnect()
```
Same as https://github.com/Open-MSS/MSS/pull/2217#discussion_r1494618935; I think this doesn't serve its intended purpose, which might be the underlying bug.

matrss commented on tests/_test_msui/test_mscolab.py:
```diff
@@ -344,6 +344,36 @@ def test_view_open(self):
+        self._activate_operation_at_index(0)
+        self.window.actionTopView.trigger()
+        QtWidgets.QApplication.processEvents()
+        assert len(self.window.get_active_views()) == 1
```
Qt is an inherently asynchronous system (at least once signals and slots come into play). This means that there is no guarantee (even with the processEvents call) that all events pertaining to creating the view are already handled once this assert is reached. I.e. there is a potential race condition between this assert and creating the view, which can fail the test.

A better approach is to use pytest-qt's `wait*` methods, specifically `wait_until`: https://pytest-qt.readthedocs.io/en/latest/wait_until.html

I've used it in other locations already to fix potential races that made the test suite unreliable, you can take a look there as well.

This applies to the other asserts as well.

rohit2p committed:
Fix: Refined QMessageBox mocking (#2220)

matrss committed:
Enable dependabot updates for GitHub Actions (#2223)

This configuration will check the GitHub Actions workflows for outdated
versions of actions and will open update PRs for those every week on
Monday (the current default for a weekly schedule).

Preetam-Das26 committed:
Return aware datetime objects (#2209)

* 15/02/2024 22:14

* 16/02/2024 14:34

* 16/02/2024 18:06

* 16/02/2024 18:06

* 16/02/2024 20:13

* 16/02/2024 20:13

* 16/02/2024 21:59

* 16/02/2024 22:53

* 20/02/2024 11:39

* 20/02/2024 14:01

dependabot[bot] committed:
Bump actions/setup-python from 4 to 5 (#2224)

Bumps [actions/setup-python](https://github.com/actions/setup-python) from 4 to 5.
- [Release notes](https://github.com/actions/setup-python/releases)
- [Commits](https://github.com/actions/setup-python/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/setup-python
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

workaryangupta committed:
Hardcoded img conventionf or route (#2218)

Preetam-Das26 committed:
Do not use datetime.datetime.utcnow() (#2225)

use always tz=datetime.timezone.utc

Myrausman committed:
Updated urls in document `NOTICE` (#2228)

Signed-off-by: Myrausman <maira.usman5703@gmail.com>

gisi90 committed:
Fix typo for flight level coordinate in netCDF4tools.py (#2235)

workaryangupta committed:
Refactor: Implement consistent usage of mscolab_settings.ARCHIVE_THRESHOLD in seed module (#2234)

matrss committed:
Remove all occurrences of processEvents in tests/ (#2237)

* Remove all occurrences of processEvents in tests/

* Add a test that ensures no test uses processEvents

matrss committed:
Remove all occurrences of qWait in tests/ (#2246)

* Remove all occurrences of qWait in tests/

* Fix some race conditions

* Add a test that ensures no test uses qWait

Myrausman committed:
refactor : Used urljoin instead of string concatenation (#2240)

Signed-off-by: Myrausman <maira.usman5703@gmail.com>

Myrausman committed:
removed section `Setup msui_settings.json for special tests` (#2252)

Preetam-Das26 committed:
Use of Aware datetime objects  (#2239)

Datetime objects are treated by many datetime methods as local times, it is preferred to use aware datetimes to represent times in UTC

matrss committed:
Fix some race conditions (#2255)

matrss committed:
Simplify GitHub Actions workflows (#2173)

This removes a bit of unnecessary code. The resulting workflows should
do effectively the same as before. The changes and reasoning are these:
- remove `success()` in `if:` blocks, since that is the implicit default
- replace `source` and `activate` calls with just `mamba run`, because
  that is more concise and readable
- remove ineffective `inputs:` definition for pull_request event,
  because it does nothing
- remove the specification of bash as the shell, since the default of
  `sh` works just fine
- rearrange shell invocations to remove unnecessary backslashes, because
  I find that to be much more pleasant to read
- replace `always()` with `!cancelled()` in `if:`, since the first would
  even run if the workflow is manually cancelled
- move trusting the git repository to the location where it is needed
- do not hardcode specific GSoC branches
- remove unused env variables

joernu76 committed:
Set data to NaN at right border of non-global data sets to avoid extrapolation issues. (#2257)

Fix #2226

ReimarBauer committed:
mscolab app does not need a static location from the gallery_builder (#2250)

ReimarBauer committed:
prepare v8.3.4 (#2259)

rohit2p committed:
Replace: occurrences of tests.utils.wait_until_signal with qtbot.wait_signal (#2247)

nilupulmanodya committed:
fix: typo (#2264)

Myrausman committed:
refactor test (#2268)

Preetam-Das26 committed:
Coverage for all branches and PRs in Coveralls (#2272)

ReimarBauer committed:
improved

ReimarBauer committed:
test improved

ReimarBauer commented:
Another test needs to verify what happens when multiple operations are drawn and then cycle login.

ReimarBauer committed:
qtbot used

ReimarBauer committed:
rebased, missing qtbot added

ReimarBauer commented:
hmm, redoing
